ChuckNorrisFactsTests/FeedViewModelTests.swift

```diff
@@ -0,0 +1,95 @@
+//
+//  FeedViewModelTests.swift
+//  ChuckNorrisFactsTests
+//
+//  Created by Tatiana Magdalena on 12/03/18.
+//  Copyright © 2018 Tatiana Magdalena. All rights reserved.
+//
+
+import XCTest
+import Nimble
+import RxBlocking
+import RxSwift
+@testable import ChuckNorrisFacts
+
+class FeedViewModelTests: XCTestCase {
+    
+    private var viewModelUnderTest: FeedViewModel!
+    private var dataSourceMock: MyMocks.FactsDataSource!
+    private var factsMapperMock: MyMocks.FactsModelMapper!
+    
+    override func setUp() {
+        super.setUp()
+    }
+    
+    override func tearDown() {
+        viewModelUnderTest = nil
+        dataSourceMock = nil
+        super.tearDown()
+    }
+    
+    func test_Successfull() {
+        prepareForTest(dataSourceResponseType: .success, apiResponseType: .validJSON)
+        
+        do {
+            let response = try viewModelUnderTest.getFactsList(searchTerm: "some-term")
+                .toBlocking()
+                .toArray()
+            expect(self.dataSourceMock.getFactsCallCount).to(equal(1))
+            expect(response.count).to(equal(2))
+            if response.count == 2 {
+                assertLoading(state: response[0], caller: #function)
+                assertSuccess(state: response[1], caller: #function)
+            }
+        } catch {
+            fail("An error was thrown when it wasn't expected to. Error: \(error)")
+        }
+        
+    }
+    
+    func test_Error() {
+        prepareForTest(dataSourceResponseType: .error, apiResponseType: .validJSON)
+        
+        do {
+            let response = try viewModelUnderTest.getFactsList(searchTerm: "some-term")
+                .toBlocking()
+                .toArray()
+            expect(self.dataSourceMock.getFactsCallCount).to(equal(1))
+            expect(response.count).to(equal(2))
+            if response.count == 2 {
+                assertLoading(state: response[0], caller: #function)
+                assertError(state: response[1], caller: #function)
+            }
+        } catch {
+            fail("View model should be able to recover from error.")
+        }
+    }
+    
+    private func prepareForTest(dataSourceResponseType: Dummy.DataSourceResponse, apiResponseType: Dummy.APIResponse) {
+        dataSourceMock = MyMocks.FactsDataSource(responseType: dataSourceResponseType)
+        factsMapperMock = MyMocks.FactsModelMapper(responseType: apiResponseType)
+        viewModelUnderTest = FeedViewModel(factsSource: dataSourceMock, factsMapper: factsMapperMock, stateMapper: FeedStateMapper())
+    }
+    
+    private func assertLoading(state: FeedState, caller: String) {
+        switch state {
+        case .loading: break
+        default: fail()
+        }
+    }
+    
+    private func assertSuccess(state: FeedState, caller: String) {
+        switch state {
+        case .success: break
```

Comment:
Switch com um case só?

Não poderia ser `guard (condition( ) ) else fail( )` ?

Repare : `condition( )` até poderia ser um lamba, e você poderia ter um método só de assertion parametrizável via esse lamba!

Reply:
O problema com o guard é que eu teria que fazer o comparativo "==", que demanda que o que eu to comparando implemente o protocolo Equatable. Que é o mesmo problema de eu fazer o `expect(value).to(equal(expectedValue)`, que também requer o Equatable.
Pra não ficar adicionando um comportamento no meu projeto, sendo que só uso ele nos testes, vi em alguns artigos recomendarem fazer desse jeito, com o switch. Dessa forma eu consigo saber qual é o case sem necessariamente fazer o "==" 🙂

Reply:
Agora, pra usar essa questão do lambda... eu vou ter que pensar um pouquinho se teria como

Reply:
Entendi. E se fosse com com `if`, precisa de `Equatable`?

if (!condition ( ) ) fail( )

Reply:
Vai dar na mesma pelo visto, então desencana, rs